### PR TITLE
Enumerize SMACK's Python Scripts

### PR DESCRIPTION
**Pull request by shaobo-he**
This PR intends to replace string literals such as SMACK's results and checks with Python's enum classes.

- [x] Result handling
- [x] `--check` option
- [x] Fix sv-comp scripts

### CODE REVIEW

**Review by zvonimir — APPROVED** [review]
I fine fine with the previous implementation too, but this also looks good to me.

**Review by keram88 — APPROVED** [review]
This looks good. I think it will be helpful in the long run to avoid our current, ad hoc approach.

**Comment by zvonimir** [comment]
Are we ok to merge this? @shaobo-he ?
SVCOMP scripts changed a lot, and so we should probably take care of them in a separate pull request.

**Comment by shaobo-he** [comment]
> Are we ok to merge this? @shaobo-he ?
> SVCOMP scripts changed a lot, and so we should probably take care of them in a separate pull request.

SVCOMP scripts won't work with the current implementation of the PR. I propose we wait until these scripts are stable.

**Comment by zvonimir** [comment]
Ok, I merged in the SCVOMP cleanup pull request. Could you now wrap this up @shaobo-he ? Thx!

File: share/smack/svcomp/utils.py [hunk 1]
```diff
@@ -48,15 +48,18 @@ def svcomp_check_property(args):
   if args.svcomp_property:
     with open(args.svcomp_property, "r") as f:
       prop = f.read()
+    from smack.top import VProperty
+    from smack.top import VResult
     if "valid-deref" in prop:
-      args.check = ['memory-safety']
+      args.check = VProperty.MEMORY_SAFETY
```
Comment (zvonimir) [inline share/smack/svcomp/utils.py:54]: Should this be a list/set?

File: share/smack/svcomp/utils.py [hunk 2]
```diff
@@ -73,7 +76,12 @@ def verify_bpl_svcomp(args):
   """Verify the Boogie source file using SVCOMP-tuned heuristics."""
   heurTrace = "\n\nHeuristics Info:\n"
 
-  if not 'memory-safety' in args.check and not 'memleak' in args.check and not 'integer-overflow' in args.check:
+  from smack.top import VProperty
+  from smack.top import VResult
+
+  if (VProperty.MEMORY_SAFETY not in args.check
```
Comment (zvonimir) [inline share/smack/svcomp/utils.py:82]: We could simplify this by pulling `not` out.

Comment (zvonimir) [reply share/smack/svcomp/utils.py:82]: You have this pattern I suggested just below.

File: share/smack/top.py [hunk 3]
```diff
@@ -446,10 +611,8 @@ def arguments():
         args.bpl_file = 'a.bpl' if args.no_verify else temporary_file(
             'a', '.bpl', args)
 
-    if not args.check:
-        args.check = {'assertions'}
-    else:
-        args.check = set([val for sublist in args.check for val in sublist])
+    if args.check == VProperty.NONE:
+        args.check = VProperty.ASSERTIONS
```
Comment (zvonimir) [inline share/smack/top.py:615]: Shouldn't this be a list/set?

**Review by zvonimir — CHANGES_REQUESTED** [review]

File: share/smack/svcomp/utils.py (hunk 1 above)
Comment (shaobo-he) [reply share/smack/svcomp/utils.py:54]: No, `args.check` is an element of class `VProperty`.

**Review by zvonimir — APPROVED** [review]
Looks good to me.

File: share/smack/svcomp/utils.py (hunk 1 above)
Comment (shaobo-he) [reply share/smack/svcomp/utils.py:54]: If I understand it correctly, an enum like this in Python is basically a bitmask. So a list of properties like what we had amounts to a mask having certain bits set.

**Pull request merged**